src/main/java/org/jetbrains/research/intellijdeodorant/ide/refactoring/ReplaceTypeCodeWithStateStrategy.java

```diff
@@ -0,0 +1,2164 @@
+                }
+            }
+            if (sourceTypeRequiredForExtraction()) {
+                methodInvocationArgumentsRewrite.add(elementFactory.createExpressionFromText("this", null));
+            }
+            PsiStatement expressionStatement = elementFactory.createStatementFromText(
+                    abstractMethodInvocation.getText() + ";",
+                    null
+            );
+            typeCheckElimination.getTypeCheckCodeFragment().replace(expressionStatement);
+        } else {
+            PsiMethodCallExpression abstractMethodInvocation = (PsiMethodCallExpression) elementFactory.createExpressionFromText(
+                    typeCheckElimination.getTypeField().getName() + "." + typeCheckElimination.getAbstractMethodName() + "();",
+                    null
+            );
+            PsiExpressionList methodInvocationArgumentsRewrite = abstractMethodInvocation.getArgumentList();
+            if (returnedVariable != null) {
+                PsiExpression argument = elementFactory.createExpressionFromText(returnedVariable.getName(), null);
+                methodInvocationArgumentsRewrite.add(argument);
+            }
+            for (PsiParameter abstractMethodParameter : typeCheckElimination.getAccessedParameters()) {
+                if (!abstractMethodParameter.equals(returnedVariable)) {
+                    PsiExpression argument = elementFactory.createExpressionFromText(abstractMethodParameter.getName(), null);
+                    methodInvocationArgumentsRewrite.add(argument);
+                }
+            }
+            for (PsiVariable fragment : typeCheckElimination.getAccessedLocalVariables()) {
+                if (!fragment.equals(returnedVariable)) {
+                    PsiExpression argument = elementFactory.createExpressionFromText(fragment.getName(), null);
+                    methodInvocationArgumentsRewrite.add(argument);
+                }
+            }
+            if (sourceTypeRequiredForExtraction()) {
+                methodInvocationArgumentsRewrite.add(elementFactory.createExpressionFromText("this", null));
+            }
+            if (returnedVariable != null) {
+                PsiStatement expressionStatement = elementFactory.createStatementFromText(
+                        returnedVariable.getName() + " = " + abstractMethodInvocation.getText() + ";",
+                        null
+                );
+                typeCheckElimination.getTypeCheckCodeFragment().replace(expressionStatement);
+            } else {
+                PsiStatement expressionStatement = elementFactory.createStatementFromText(
+                        "return " + abstractMethodInvocation.getText() + ";",
+                        null
+                );
+                typeCheckElimination.getTypeCheckCodeFragment().replace(expressionStatement);
+            }
+        }
+    }
+
+    //
+//	private void initializeReturnedVariableDeclaration() { // TODO: do we really need it???
```

Comment:
?

Reply:
Initially I could not imagine a situation when this code could be useful. But I figured it out now and added this logic with a test for it.